File: todo.java
// todo.java
import java.io.*;
import java.nio.file.*;
import java.time.*;
import java.util.*;
import com.google.gson.*; // install gson

public class todo {
    private static final String RESET = "\u001B[0m";
    private static final String RED = "\u001B[91m";
    private static final String GREEN = "\u001B[92m";
    private static final String YELLOW = "\u001B[93m";
    private static final String BLUE = "\u001B[94m";
    private static final String BOLD = "\u001B[1m";

    private static String colorize(String text, String color) {
        return color + text + RESET;
    }

    private static String tasksFile = System.getProperty("user.home") + "/.todo_tasks.json";

    static class Task {
        int id;
        String title;
        String description;
        String priority;
        String deadline;
        String status;
        String created_at;
    }

    private static List<Task> loadTasks() throws IOException {
        Path path = Paths.get(tasksFile);
        if (!Files.exists(path)) return new ArrayList<>();
        String json = new String(Files.readAllBytes(path));
        Gson gson = new Gson();
        Task[] arr = gson.fromJson(json, Task[].class);
        return new ArrayList<>(Arrays.asList(arr));
    }

    private static void saveTasks(List<Task> tasks) throws IOException {
        Gson gson = new GsonBuilder().setPrettyPrinting().create();
        String json = gson.toJson(tasks);
        Files.write(Paths.get(tasksFile), json.getBytes());
    }

    private static int getNextID(List<Task> tasks) {
        int max = 0;
        for (Task t : tasks) {
            if (t.id > max) max = t.id;
        }
        return max + 1;
    }

    private static String taskColor(Task t) {
        if (t.status.equals("done")) return GREEN;
        switch (t.priority) {
            case "High": return RED;
            case "Medium": return YELLOW;
            case "Low": return BLUE;
            default: return RESET;
        }
    }

    private static String prioritySymbol(Task t) {
        switch (t.priority) {
            case "High": return "🔴";
            case "Medium": return "🟡";
            case "Low": return "🔵";
            default: return "⚪";
        }
    }

    private static void addTask(String title, String description, String priority, String deadline) throws IOException {
        List<Task> tasks = loadTasks();
        Task t = new Task();
        t.id = getNextID(tasks);
        t.title = title;
        t.description = description == null ? "" : description;
        t.priority = priority == null ? "Medium" : priority;
        t.deadline = deadline == null ? "" : deadline;
        t.status = "pending";
        t.created_at = Instant.now().toString();
        tasks.add(t);
        saveTasks(tasks);
        System.out.println(colorize("✅ Задача #" + t.id + " добавлена.", GREEN));
    }

    private static void listTasks(String filter, String sortBy) throws IOException {
        List<Task> tasks = loadTasks();
        if (tasks.isEmpty()) {
            System.out.println(colorize("Нет задач.", YELLOW));
            return;
        }
        if (filter != null) {
            String[] parts = filter.split("=");
            if (parts.length == 2) {
                String key = parts[0], val = parts[1];
                List<Task> filtered = new ArrayList<>();
                for (Task t : tasks) {
                    if (key.equals("status") && t.status.equals(val)) filtered.add(t);
                    else if (key.equals("priority") && t.priority.equals(val)) filtered.add(t);
                }
                tasks = filtered;
            }
        }
        if (sortBy != null) {
            switch (sortBy) {
                case "priority":
                    Map<String, Integer> order = new HashMap<>();
                    order.put("High", 0);
                    order.put("Medium", 1);
                    order.put("Low", 2);
                    tasks.sort((a, b) -> order.get(a.priority) - order.get(b.priority));
                    break;
                case "deadline":
                    tasks.sort((a, b) -> {
                        String da = a.deadline.isEmpty() ? "9999-12-31" : a.deadline;
                        String db = b.deadline.isEmpty() ? "9999-12-31" : b.deadline;
                        return da.compareTo(db);
                    });
                    break;
                case "created":
                    tasks.sort((a, b) -> a.created_at.compareTo(b.created_at));
                    break;
            }
        }
        for (Task t : tasks) {
            String status = t.status.equals("done") ? "✅" : "⬜";
            String deadline = t.deadline.isEmpty() ? "" : " до " + t.deadline;
            System.out.printf("[%d] %s %s %s %s\n",
                t.id, status, t.title, colorize(t.priority, taskColor(t)), deadline);
        }
    }

    private static void doneTask(int id) throws IOException {
        List<Task> tasks = loadTasks();
        Task found = null;
        for (Task t : tasks) {
            if (t.id == id) { found = t; break; }
        }
        if (found == null) {
            System.out.println(colorize("Задача #" + id + " не найдена.", RED));
            return;
        }
        if (found.status.equals("done")) {
            System.out.println(colorize("Задача #" + id + " уже выполнена.", YELLOW));
            return;
        }
        found.status = "done";
        saveTasks(tasks);
        System.out.println(colorize("✅ Задача #" + id + " отмечена как выполненная.", GREEN));
    }

    private static void deleteTask(int id) throws IOException {
        List<Task> tasks = loadTasks();
        Iterator<Task> it = tasks.iterator();
        while (it.hasNext()) {
            Task t = it.next();
            if (t.id == id) {
                it.remove();
                saveTasks(tasks);
                System.out.println(colorize("🗑️ Задача #" + id + " удалена.", YELLOW));
                return;
            }
        }
        System.out.println(colorize("Задача #" + id + " не найдена.", RED));
    }

    private static void editTask(int id, String title, String description, String priority, String deadline) throws IOException {
        List<Task> tasks = loadTasks();
        Task found = null;
        for (Task t : tasks) {
            if (t.id == id) { found = t; break; }
        }
        if (found == null) {
            System.out.println(colorize("Задача #" + id + " не найдена.", RED));
            return;
        }
        if (title != null) found.title = title;
        if (description != null) found.description = description;
        if (priority != null) found.priority = priority;
        if (deadline != null) found.deadline = deadline;
        saveTasks(tasks);
        System.out.println(colorize("✅ Задача #" + id + " обновлена.", GREEN));
    }

    private static void searchTasks(String query) throws IOException {
        List<Task> tasks = loadTasks();
        List<Task> results = new ArrayList<>();
        String q = query.toLowerCase();
        for (Task t : tasks) {
            if (t.title.toLowerCase().contains(q) || t.description.toLowerCase().contains(q)) {
                results.add(t);
            }
        }
        if (results.isEmpty()) {
            System.out.println(colorize("Ничего не найдено по запросу '" + query + "'.", YELLOW));
            return;
        }
        System.out.println(colorize("🔍 Найдено " + results.size() + " задач:", BOLD));
        for (Task t : results) {
            System.out.println("  [" + t.id + "] " + t.title);
        }
    }

    private static void stats() throws IOException {
        List<Task> tasks = loadTasks();
        int total = tasks.size();
        int done = 0;
        for (Task t : tasks) {
            if (t.status.equals("done")) done++;
        }
        int pending = total - done;
        System.out.println(colorize("📊 Статистика:", BOLD));
        System.out.println("  Всего задач: " + total);
        if (total > 0) {
            System.out.printf("  Выполнено: %d (%.1f%%)\n", done, (double)done / total * 100);
        } else {
            System.out.println("  Выполнено: 0");
        }
        System.out.println("  Ожидают: " + pending);
    }

    private static void exportTasks(String filename) throws IOException {
        List<Task> tasks = loadTasks();
        Gson gson = new GsonBuilder().setPrettyPrinting().create();
        String json = gson.toJson(tasks);
        Files.write(Paths.get(filename), json.getBytes());
        System.out.println(colorize("✅ Задачи экспортированы в " + filename, GREEN));
    }

    private static void importTasks(String filename) throws IOException {
        if (!Files.exists(Paths.get(filename))) {
            System.out.println(colorize("Файл не найден.", RED));
            return;
        }
        String json = new String(Files.readAllBytes(Paths.get(filename)));
        Gson gson = new Gson();
        Task[] arr = gson.fromJson(json, Task[].class);
        List<Task> tasks = loadTasks();
        for (Task t : arr) {
            t.id = getNextID(tasks);
            tasks.add(t);
        }
        saveTasks(tasks);
        System.out.println(colorize("✅ Импортировано " + arr.length + " задач из " + filename, GREEN));
    }

    public static void main(String[] args) throws IOException {
        if (args.length == 0) {
            System.out.println("Usage: java todo <add|list|done|delete|edit|search|stats|export|import> [args...]");
            System.out.println("  add <title> [description] -p <priority> -d <deadline>");
            System.out.println("  list [--filter status=pending|done] [--sort priority|deadline|created]");
            System.out.println("  done <id>");
            System.out.println("  delete <id>");
            System.out.println("  edit <id> --title <title> --description <desc> --priority <p> --deadline <d>");
            System.out.println("  search <query>");
            System.out.println("  stats");
            System.out.println("  export <file>");
            System.out.println("  import <file>");
            return;
        }
        String cmd = args[0];

        switch (cmd) {
            case "add": {
                if (args.length < 2) {
                    System.out.println("Использование: add <title> [description] -p <priority> -d <deadline>");
                    return;
                }
                String title = args[1];
                String description = null, priority = "Medium", deadline = null;
                for (int i = 2; i < args.length; i++) {
                    if (args[i].equals("-p") && i+1 < args.length) {
                        priority = args[++i];
                    } else if (args[i].equals("-d") && i+1 < args.length) {
                        deadline = args[++i];
                    } else if (description == null) {
                        description = args[i];
                    }
                }
                addTask(title, description, priority, deadline);
                break;
            }
            case "list": {
                String filter = null, sortBy = null;
                for (int i = 1; i < args.length; i++) {
                    if (args[i].equals("--filter") && i+1 < args.length) {
                        filter = args[++i];
                    } else if (args[i].equals("--sort") && i+1 < args.length) {
                        sortBy = args[++i];
                    }
                }
                listTasks(filter, sortBy);
                break;
            }
            case "done": {
                if (args.length < 2) {
                    System.out.println("Использование: done <id>");
                    return;
                }
                doneTask(Integer.parseInt(args[1]));
                break;
            }
            case "delete": {
                if (args.length < 2) {
                    System.out.println("Использование: delete <id>");
                    return;
                }
                deleteTask(Integer.parseInt(args[1]));
                break;
            }
            case "edit": {
                if (args.length < 3) {
                    System.out.println("Использование: edit <id> --title <title> --description <desc> --priority <p> --deadline <d>");
                    return;
                }
                int id = Integer.parseInt(args[1]);
                String title = null, description = null, priority = null, deadline = null;
                for (int i = 2; i < args.length; i++) {
                    if (args[i].equals("--title") && i+1 < args.length) {
                        title = args[++i];
                    } else if (args[i].equals("--description") && i+1 < args.length) {
                        description = args[++i];
                    } else if (args[i].equals("--priority") && i+1 < args.length) {
                        priority = args[++i];
                    } else if (args[i].equals("--deadline") && i+1 < args.length) {
                        deadline = args[++i];
                    }
                }
                editTask(id, title, description, priority, deadline);
                break;
            }
            case "search": {
                if (args.length < 2) {
                    System.out.println("Использование: search <query>");
                    return;
                }
                searchTasks(args[1]);
                break;
            }
            case "stats":
                stats();
                break;
            case "export": {
                if (args.length < 2) {
                    System.out.println("Использование: export <file>");
                    return;
                }
                exportTasks(args[1]);
                break;
            }
            case "import": {
                if (args.length < 2) {
                    System.out.println("Использование: import <file>");
                    return;
                }
                importTasks(args[1]);
                break;
            }
            default:
                System.out.println("Неизвестная команда: " + cmd);
        }
    }
}

File: todo.cs
// todo.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Text.Json.Serialization;

class Task
{
    public int Id { get; set; }
    public string Title { get; set; }
    public string Description { get; set; }
    public string Priority { get; set; }
    public string Deadline { get; set; }
    public string Status { get; set; }
    public string CreatedAt { get; set; }
}

class Todo
{
    static string Colorize(string text, string color)
    {
        string col = color switch
        {
            "red" => "\x1b[91m",
            "green" => "\x1b[92m",
            "yellow" => "\x1b[93m",
            "blue" => "\x1b[94m",
            "bold" => "\x1b[1m",
            _ => "\x1b[0m"
        };
        return col + text + "\x1b[0m";
    }

    static string TasksFile = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), ".todo_tasks.json");

    static List<Task> LoadTasks()
    {
        if (!File.Exists(TasksFile)) return new List<Task>();
        string json = File.ReadAllText(TasksFile);
        return JsonSerializer.Deserialize<List<Task>>(json) ?? new List<Task>();
    }

    static void SaveTasks(List<Task> tasks)
    {
        string json = JsonSerializer.Serialize(tasks, new JsonSerializerOptions { WriteIndented = true });
        File.WriteAllText(TasksFile, json);
    }

    static int GetNextID(List<Task> tasks) => tasks.Count > 0 ? tasks.Max(t => t.Id) + 1 : 1;

    static string TaskColor(Task t)
    {
        if (t.Status == "done") return "green";
        return t.Priority switch
        {
            "High" => "red",
            "Medium" => "yellow",
            "Low" => "blue",
            _ => "reset"
        };
    }

    static string PrioritySymbol(Task t)
    {
        return t.Priority switch
        {
            "High" => "🔴",
            "Medium" => "🟡",
            "Low" => "🔵",
            _ => "⚪"
        };
    }

    static void AddTask(string title, string description, string priority, string deadline)
    {
        var tasks = LoadTasks();
        var task = new Task
        {
            Id = GetNextID(tasks),
            Title = title,
            Description = description ?? "",
            Priority = priority ?? "Medium",
            Deadline = deadline ?? "",
            Status = "pending",
            CreatedAt = DateTime.Now.ToString("o")
        };
        tasks.Add(task);
        SaveTasks(tasks);
        Console.WriteLine(Colorize($"✅ Задача #{task.Id} добавлена.", "green"));
    }

    static void ListTasks(string filter, string sortBy)
    {
        var tasks = LoadTasks();
        if (tasks.Count == 0)
        {
            Console.WriteLine(Colorize("Нет задач.", "yellow"));
            return;
        }
        if (!string.IsNullOrEmpty(filter))
        {
            var parts = filter.Split('=');
            if (parts.Length == 2)
            {
                string key = parts[0], val = parts[1];
                tasks = tasks.Where(t =>
                    (key == "status" && t.Status == val) ||
                    (key == "priority" && t.Priority == val)
                ).ToList();
            }
        }
        if (sortBy == "priority")
        {
            var order = new Dictionary<string, int> { { "High", 0 }, { "Medium", 1 }, { "Low", 2 } };
            tasks = tasks.OrderBy(t => order[t.Priority]).ToList();
        }
        else if (sortBy == "deadline")
        {
            tasks = tasks.OrderBy(t => t.Deadline ?? "9999-12-31").ToList();
        }
        else if (sortBy == "created")
        {
            tasks = tasks.OrderBy(t => t.CreatedAt).ToList();
        }
        foreach (var t in tasks)
        {
            string status = t.Status == "done" ? "✅" : "⬜";
            string deadline = string.IsNullOrEmpty(t.Deadline) ? "" : $" до {t.Deadline}";
            Console.WriteLine($"[{t.Id}] {status} {t.Title} {Colorize(t.Priority, TaskColor(t))} {deadline}");
        }
    }

    static void DoneTask(int id)
    {
        var tasks = LoadTasks();
        var task = tasks.FirstOrDefault(t => t.Id == id);
        if (task == null)
        {
            Console.WriteLine(Colorize($"Задача #{id} не найдена.", "red"));
            return;
        }
        if (task.Status == "done")
        {
            Console.WriteLine(Colorize($"Задача #{id} уже выполнена.", "yellow"));
            return;
        }
        task.Status = "done";
        SaveTasks(tasks);
        Console.WriteLine(Colorize($"✅ Задача #{id} отмечена как выполненная.", "green"));
    }

    static void DeleteTask(int id)
    {
        var tasks = LoadTasks();
        var idx = tasks.FindIndex(t => t.Id == id);
        if (idx == -1)
        {
            Console.WriteLine(Colorize($"Задача #{id} не найдена.", "red"));
            return;
        }
        tasks.RemoveAt(idx);
        SaveTasks(tasks);
        Console.WriteLine(Colorize($"🗑️ Задача #{id} удалена.", "yellow"));
    }

    static void EditTask(int id, string title, string description, string priority, string deadline)
    {
        var tasks = LoadTasks();
        var task = tasks.FirstOrDefault(t => t.Id == id);
        if (task == null)
        {
            Console.WriteLine(Colorize($"Задача #{id} не найдена.", "red"));
            return;
        }
        if (!string.IsNullOrEmpty(title)) task.Title = title;
        if (!string.IsNullOrEmpty(description)) task.Description = description;
        if (!string.IsNullOrEmpty(priority)) task.Priority = priority;
        if (!string.IsNullOrEmpty(deadline)) task.Deadline = deadline;
        SaveTasks(tasks);
        Console.WriteLine(Colorize($"✅ Задача #{id} обновлена.", "green"));
    }

    static void SearchTasks(string query)
    {
        var tasks = LoadTasks();
        var results = tasks.Where(t =>
            t.Title.Contains(query, StringComparison.OrdinalIgnoreCase) ||
            t.Description.Contains(query, StringComparison.OrdinalIgnoreCase)
        ).ToList();
        if (results.Count == 0)
        {
            Console.WriteLine(Colorize($"Ничего не найдено по запросу '{query}'.", "yellow"));
            return;
        }
        Console.WriteLine(Colorize($"🔍 Найдено {results.Count} задач:", "bold"));
        foreach (var t in results) Console.WriteLine($"  [{t.Id}] {t.Title}");
    }

    static void Stats()
    {
        var tasks = LoadTasks();
        int total = tasks.Count;
        int done = tasks.Count(t => t.Status == "done");
        int pending = total - done;
        Console.WriteLine(Colorize("📊 Статистика:", "bold"));
        Console.WriteLine($"  Всего задач: {total}");
        if (total > 0)
            Console.WriteLine($"  Выполнено: {done} ({(double)done/total*100:F1}%)");
        else
            Console.WriteLine("  Выполнено: 0");
        Console.WriteLine($"  Ожидают: {pending}");
    }

    static void ExportTasks(string filename)
    {
        var tasks = LoadTasks();
        string json = JsonSerializer.Serialize(tasks, new JsonSerializerOptions { WriteIndented = true });
        File.WriteAllText(filename, json);
        Console.WriteLine(Colorize($"✅ Задачи экспортированы в {filename}", "green"));
    }

    static void ImportTasks(string filename)
    {
        if (!File.Exists(filename))
        {
            Console.WriteLine(Colorize("Файл не найден.", "red"));
            return;
        }
        string json = File.ReadAllText(filename);
        var imported = JsonSerializer.Deserialize<List<Task>>(json) ?? new List<Task>();
        var tasks = LoadTasks();
        foreach (var t in imported)
        {
            t.Id = GetNextID(tasks);
            tasks.Add(t);
        }
        SaveTasks(tasks);
        Console.WriteLine(Colorize($"✅ Импортировано {imported.Count} задач из {filename}", "green"));
    }

    static void Main(string[] args)
    {
        if (args.Length == 0)
        {
            Console.WriteLine("Usage: todo <add|list|done|delete|edit|search|stats|export|import> [args...]");
            return;
        }
        string cmd = args[0];

        switch (cmd)
        {
            case "add":
                if (args.Length < 2) { Console.WriteLine("Использование: add <title> [description] -p <priority> -d <deadline>"); return; }
                string title = args[1], description = "", priority = "Medium", deadline = "";
                for (int i = 2; i < args.Length; i++)
                {
                    if (args[i] == "-p" && i+1 < args.Length) priority = args[++i];
                    else if (args[i] == "-d" && i+1 < args.Length) deadline = args[++i];
                    else if (string.IsNullOrEmpty(description)) description = args[i];
                }
                AddTask(title, description, priority, deadline);
                break;
            case "list":
                string filter = "", sortBy = "";
                for (int i = 1; i < args.Length; i++)
                {
                    if (args[i] == "--filter" && i+1 < args.Length) filter = args[++i];
                    else if (args[i] == "--sort" && i+1 < args.Length) sortBy = args[++i];
                }
                ListTasks(filter, sortBy);
                break;
            case "done":
                if (args.Length < 2) { Console.WriteLine("Использование: done <id>"); return; }
                DoneTask(int.Parse(args[1]));
                break;
            case "delete":
                if (args.Length < 2) { Console.WriteLine("Использование: delete <id>"); return; }
                DeleteTask(int.Parse(args[1]));
                break;
            case "edit":
                if (args.Length < 3) { Console.WriteLine("Использование: edit <id> [--title <title>] [--description <desc>] [--priority <p>] [--deadline <d>]"); return; }
                int id = int.Parse(args[1]);
                string ttl = "", desc = "", pri = "", ddl = "";
                for (int i = 2; i < args.Length; i++)
                {
                    if (args[i] == "--title" && i+1 < args.Length) ttl = args[++i];
                    else if (args[i] == "--description" && i+1 < args.Length) desc = args[++i];
                    else if (args[i] == "--priority" && i+1 < args.Length) pri = args[++i];
                    else if (args[i] == "--deadline" && i+1 < args.Length) ddl = args[++i];
                }
                EditTask(id, ttl, desc, pri, ddl);
                break;
            case "search":
                if (args.Length < 2) { Console.WriteLine("Использование: search <query>"); return; }
                SearchTasks(args[1]);
                break;
            case "stats":
                Stats();
                break;
            case "export":
                if (args.Length < 2) { Console.WriteLine("Использование: export <file>"); return; }
                ExportTasks(args[1]);
                break;
            case "import":
                if (args.Length < 2) { Console.WriteLine("Использование: import <file>"); return; }
                ImportTasks(args[1]);
                break;
            default:
                Console.WriteLine($"Неизвестная команда: {cmd}");
                break;
        }
    }
}
